fix(http): allow URL parameters to be encoded

tiangolo opened the pull request:
Back ends for several languages (including Python, PHP and Node.js) expect to receive encoded parameters with some standard conventions, when they receive them as raw decoded strings they can't parse those parameters. This removes the custom decoding of parameters to raw strings and leaves the default standard encoding for HttpClient.

## PR Checklist
Please check if your PR fulfills the following requirements:

- [ :heavy_check_mark:  ] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [  ] Tests for the changes have been added (for bug fixes / features)
> Please check the comments to discuss if this is a change you are willing to accept.
- [ N/A ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->
```
[ x ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] angular.io application / infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Currently, when creating URL parameters, some characters are decoded and converted to raw strings after the `encodeURIComponent()` had encoded them. The characters are: `@`, `:`, `$`, `,`, `;`, `+`, `=`, `?`, `/`.

This creates communication problems with back ends in at least Python, PHP and Node.js.

Check the referenced issues for more details on all the possible problems this creates.

Issue Number: #11058 #18884 #18274 #14531 #13077 #18261


## What is the new behavior?

With this PR, when using `URLSearchParams`, the parameters will be encoded as normally, without being decoded by Angular before sending them in the URL to the server.

I created a parallel PR that solves the problem for cases using `HttpParams` here: https://github.com/angular/angular/pull/19710.

## Does this PR introduce a breaking change?
```
[ ] Yes
[ x ] No
```
...or would you consider this a breaking change?

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

tiangolo commented:
Before modifying the tests, I would like to discuss it here.

I see there are references in the code to https://tools.ietf.org/html/rfc3986#section-3.4 and https://github.com/angular/angular/issues/9348

Specifically:

```
       * https://tools.ietf.org/html/rfc3986#section-3.4
       * Allowed: ( pchar / "/" / "?" )
       * pchar: unreserved / pct-encoded / sub-delims / ":" / "@"
       * unreserved: ALPHA / DIGIT / "-" / "." / "_" / "~"
       * pct-encoded: "%" HEXDIG HEXDIG
       * sub-delims: "!" / "$" / "&" / "'" / "(" / ")" / "*" / "+" / "," / ";" / "="
       *
       * & and = are excluded and should be encoded inside keys and values
       * because URLSearchParams is responsible for inserting this.
```

But I see the RFC is referring to the whole query component, everything after the `?`. So, as I see it, it describes all the allowed characters, including the separation characters. I don't see it saying something like that query "parameters" inside the "query component" that contain reserved characters should not be escaped.

Actually, I don't see the RFC impose or describe how to use the query section, apart from describing all the allowed characters. I only see it mention an example of how to use and structure the query component with an "often used":

> [...] URI producing applications often use the reserved characters allowed in a segment to delimit scheme-specific or dereference-handler-specific subcomponents.  For example, the semicolon (";") and equals ("=") reserved characters are often used to delimit parameters and parameter values applicable to that segment. [...]

But the `URLSearchParams` is not encoding a full query component but encoding query component parameters, that's its purpose. And "parameters" in the query component seem to be out of the scope of that RFC, as it only describes the query component as a whole.

On the other side, I see [URL recommendations from the W3C](https://www.w3.org/Addressing/URL/4_URI_Recommentations.html) including that:

> Within the query string, the plus sign is reserved as shorthand notation for a space. Therefore, real plus signs must be encoded.

I see that in [RFC 1738](https://tools.ietf.org/html/rfc1738.html) it says that:

> [...] Within the <path> and <searchpart> components, "/", ";", "?" are
   reserved. [...]

Apart from that, I don't see any authoritative official reference on how to use the query component more specifically.

The [WHATWG URL spec](https://url.spec.whatwg.org/) doesn't seem to have anything too explicit about how to encode the query component.

I see in the [Python docs](https://docs.python.org/3.6/library/urllib.parse.html?highlight=query%20component#urllib.parse.parse_qs) that it is used as a string encoded with `application/x-www-form-urlencoded`.

In [Node.js docs](https://nodejs.org/api/querystring.html#querystring_querystring_parse_str_sep_eq_options) it doesn't mention a specific encoding, but it says that JS's `decodeURIComponent` is used (the contrary to `encodeURIComponent`).

[PHP docs](http://php.net/manual/en/function.urldecode.php) say they decode percent encoded strings. And convert plus symbols (`+`) to spaces.

The [W3C HTML spec](https://www.w3.org/TR/html401/interact/forms.html#h-17.13.4.1) describes how to encode with `application/x-www-form-urlencoded`, and states that:

> 1. Control names and values are escaped. Space characters are replaced by `+`, and then reserved characters are escaped as described in [RFC1738], section 2.2: Non-alphanumeric characters are replaced by `%HH`, a percent sign and two hexadecimal digits representing the ASCII code of the character. Line breaks are represented as "CR LF" pairs (i.e., `%0D%0A`).
> 2. The control names/values are listed in the order they appear in the document. The name is separated from the value by `=` and name/value pairs are separated from each other by `&`.

It seems to me like the most complete description of how to use the query component would be the one from the W3C, although it actually referes to HTTP POST data, but is what is used by Python. And other languages, although not explicitly, seem to be compatible with that implementation.

I see the solution would be to allow the `encodeURIComponent` do its thing.

Let me know what you think and I will modify the tests if this seems something you would like to have in Angular.

Splaktar commented:
Please rebase and resolve merge conflicts. Sorry for the long delay in reviewing/merging this.

tiangolo committed:
fix(http): allow URL parameters to be encoded

Back ends for several languages (including Python, PHP and Node.js) expect to receive encoded parameters with some standard conventions, when they receive them as raw decoded strings they can't parse those parameters. This removes the custom decoding of parameters to raw strings and leaves the default standard encoding for URLSearchParams.

tiangolo commented:
Thanks for the review @Splaktar , done :heavy_check_mark:

mhevery commented:
It seems that that code is intentional per this comment: https://github.com/angular/angular/blob/6046419f6ccc7a74cdd43fa22a55a405bd229698/packages/common/upgrade/src/params.ts#L316-L337
